### ISSUE TITLE
[xpu-alignment] [26-09-21] adaptive_max_pool3d_backward on XPU reads a non-contiguous indices tensor as if it were contiguous: silently wrong gradients under torch.compile with channels_last_3d inputs

### ISSUE THREAD

**Issue by torchxpubot** [intel/torch-xpu-ops#5475]
## Summary

`adaptive_max_pool3d_backward_kernel` in `src/ATen/native/xpu/sycl/AdaptiveMaxPooling3dKernels.cpp` makes `gradOutput` contiguous but reads `indices` through its raw data pointer. When `indices` arrives with `channels_last_3d` strides the kernel scatters the gradient to the wrong positions, with no error. Inductor lays the saved indices out in the input's memory format, so `F.adaptive_max_pool3d` under `torch.compile` on a `channels_last_3d` XPU input trains with silently wrong gradients through every adaptive max pool.

This is the XPU half of upstream `pytorch/pytorch#197888`, which reports the same class of defect for the CPU kernels of `adaptive_max_pool3d_backward`, `fractional_max_pool2d_backward` and `fractional_max_pool3d_backward`. Only `adaptive_max_pool3d_backward` reproduces on XPU; the two XPU `fractional_max_pool*_backward` kernels use `PackedTensorAccessor64`, honour the real strides, and are correct.

## Upstream source and scan window

- Upstream issue: https://github.com/pytorch/pytorch/issues/197888 (OPEN, no fix PR as of this scan)
- Scan window: `[2026-09-21T00:00:00Z, 2026-09-22T00:00:00Z)` on `pytorch/pytorch`
- Frozen `pytorch/pytorch` default-branch head: `6b5c7ad181277ba21aa48b1dc03786ab95f68ad4`
- Upstream oracle: the correct gradient is the one obtained when the same index values are read with their real strides (equivalently, from a contiguous copy of `indices`). Upstream reports that on CUDA the compiled gradient matches eager, so upstream's stated fix scope is the CPU kernels.

## Root cause in this repository

`src/ATen/native/xpu/sycl/AdaptiveMaxPooling3dKernels.cpp`, `adaptive_max_pool3d_backward_kernel`:

```cpp
void adaptive_max_pool3d_backward_kernel(
    const Tensor& gradOutput,
    const Tensor& input,
    const Tensor& indices,
    const Tensor& gradInput) {
  const Tensor gradOutput_ = gradOutput.contiguous();   // indices is not normalized
  ...
          const int64_t* indices_data = indices.const_data_ptr<int64_t>();
```

`gradOutput` is normalized, `indices` is not, and both templates then index `indices_data` with contiguous arithmetic. The XPU kernel is reached through `adaptive_max_pool3d_backward_out_xpu` in `src/ATen/native/xpu/AdaptiveMaxPooling3d.cpp`.

The upstream CUDA kernel (`aten/src/ATen/native/cuda/AdaptiveMaxPooling3d.cu` at the frozen head) carries the same raw-pointer pattern, but upstream reports its compiled path does not hit it; the XPU compiled path does.

## Reproducer

```python
import torch
import torch.nn.functional as F

A = torch.ops.aten

# 1. Kernel level: identical index values, contiguous vs channels_last_3d strides.
torch.manual_seed(0)
x3 = torch.randperm(240).float().reshape(1, 4, 4, 3, 5).to("xpu")
v, i = F.adaptive_max_pool3d(x3, (2, 3, 5), return_indices=True)
go = torch.arange(1.0, v.numel() + 1, device="xpu").reshape(v.shape)
i_nc = i.contiguous(memory_format=torch.channels_last_3d)  # same values, different strides
print("indices strides", tuple(i_nc.stride()))
print(
    "same gradient as with contiguous indices:",
    torch.equal(
        A.adaptive_max_pool3d_backward(go, x3, i),
        A.adaptive_max_pool3d_backward(go, x3, i_nc),
    ),
)

# 2. User level: eager vs aot_eager vs inductor on a channels_last_3d input.
torch.manual_seed(0)
f = lambda t: F.adaptive_max_pool3d(t, (2, 3, 5))
x = torch.randperm(1 * 4 * 4 * 6 * 5).float().reshape(1, 4, 4, 6, 5).to("xpu")
for name, xin in (
    ("contiguous", x),
    ("channels_last_3d", x.contiguous(memory_format=torch.channels_last_3d)),
):
    grads = []
    for backend in (None, "aot_eager", "inductor"):
        torch._dynamo.reset()
        t = xin.detach().clone(memory_format=torch.preserve_format).requires_grad_()
        out = (f if backend is None else torch.compile(f, backend=backend))(t)
        w = torch.arange(1.0, out.numel() + 1, device="xpu").reshape(out.shape)
        (out * w).sum().backward()
        grads.append(t.grad.clone())
    print(
        f"{name:17s} aot_eager==eager: {torch.equal(grads[0], grads[1])}"
        f"   inductor==eager: {torch.equal(grads[0], grads[2])}"
        f"   max diff {(grads[0] - grads[2]).abs().max().item():.0f}"
    )
```

## Observed XPU behavior

```
KERNEL: adaptive_max_pool3d_backward indices_stride=(120, 1, 60, 20, 4) same_gradient_as_contiguous_indices=False
KERNEL: adaptive_max_pool2d_backward indices_stride=(60, 1, 20, 4) same_gradient_as_contiguous_indices=True
KERNEL: max_pool3d_with_indices_backward indices_stride=(96, 1, 32, 16, 4) same_gradient_as_contiguous_indices=True
KERNEL: fractional_max_pool2d_backward indices_stride=(36, 1, 12, 4) same_gradient_as_contiguous_indices=True
KERNEL: fractional_max_pool3d_backward indices_stride=(32, 1, 16, 8, 4) same_gradient_as_contiguous_indices=True
COMPILE: contiguous aot_eager==eager=True inductor==eager=True max_diff=0.0
COMPILE: channels_last_3d aot_eager==eager=True inductor==eager=False max_diff=118.0
```

## Target-path evidence

All five pooling backward ops were invoked with `xpu` tensors and asserted to have produced `xpu` outputs, so the divergence comes from the XPU SYCL kernels rather than from a host fallback. At user level, `aot_eager==eager=True` while `inductor==eager=False` on the same `channels_last_3d` XPU input, and `t.grad.device.type` was asserted to be `xpu`; that isolates the wrong gradient to the XPU backward kernel reached through inductor's memory-format choice, not to the tracing layer.

## Environment

```
torch            2.15.0.dev20260920+xpu
torch path       /opt/conda/lib/python3.13/site-packages/torch/__init__.py
python           3.13.13 (/opt/conda/bin/python3.13)
xpu available    True
xpu device       Intel(R) Arc(TM) Pro B60 Graphics
```

## Ownership rationale

The wrong code is `adaptive_max_pool3d_backward_kernel` in this repository, reached through `adaptive_max_pool3d_backward_out_xpu`; the fix is to normalize `indices` next to the existing `gradOutput.contiguous()` (or to index it through its real strides). Upstream `pytorch/pytorch#197888` reports CUDA's compiled gradient as correct and scopes its suggested fix to the CPU kernels, so landing it would not change the XPU kernel. No `pytorch/pytorch` issue or PR was found that explicitly owns the XPU part, and no existing `intel/torch-xpu-ops` tracker covers it.

Scope note for whoever picks this up: only `adaptive_max_pool3d_backward` needs the change on XPU. `adaptive_max_pool2d_backward`, `max_pool3d_with_indices_backward` and both `fractional_max_pool*_backward` kernels already handle non-contiguous `indices` correctly, so the two `fractional_max_pool*` thirds of the upstream oracle do not apply here.

Filed by the XPU alignment automation; the finding was produced by a scan and independently reviewed before filing.

**Comment by laifenxiawucha** [comment]
@torchxpubot fix

**Comment by torchxpubot** [comment]
Automated fix session for this issue. Stages are appended below as they complete.



<details>
<summary><b>Issue Triage</b></summary>

| Field | Value |
|-------|-------|
| Issue type | single-bug |
| Reproduction missing | no |
| Scope | torch-xpu-ops |
| Dependencies | (none) |
| Handling | agent-fixable |

*Automated by issue-triage.*

</details>



<details>
<summary><b>Reproduce</b></summary>

Reproduced against the latest XPU nightly wheel `2.15.0.dev20260920+xpu` on an
Intel(R) Arc(TM) Pro B60 Graphics device, using the reproducer from the issue
body plus one extra case that forces the `atomic` variant of the same kernel
(`isizeH % osizeH != 0`).

| Test case | Verdict | Observed |
|---|---|---|
| kernel, non-atomic path (`1x4x4x3x5 -> 2x3x5`) | REPRODUCED | `same_gradient_as_contiguous_indices=False` for `indices.stride()==(120, 1, 60, 20, 4)` |
| kernel, atomic path (`1x4x4x5x5 -> 2x3x5`) | REPRODUCED | `same_gradient_as_contiguous_indices=False`, so both kernel templates are affected |
| `torch.compile` end-to-end, `channels_last_3d` input | REPRODUCED | `inductor==eager=False`, `max_diff=118`; `aot_eager==eager=True` |

`torch.compile` on a contiguous input is unaffected (`inductor==eager=True`),
which matches the issue's claim that the trigger is the memory format inductor
picks for the saved `indices`.

*Automated by fix-reproduce.*

</details>



<details>
<summary><b>Root Cause</b></summary>

Analyzed at `torch-xpu-ops@6a2d048e`, pytorch `origin/main@00ec4ca3a9`.

`adaptive_max_pool3d_backward_kernel`
(`src/ATen/native/xpu/sycl/AdaptiveMaxPooling3dKernels.cpp:500`) normalizes only
one of the three tensors its kernels address with contiguous index arithmetic.
Line 505 takes `gradOutput.contiguous()`, but `indices` is passed straight
through as a raw pointer (lines 553 and 577) and both kernel templates then walk
it as if it were contiguous:
`indices_ + o_plane * osizeH_ * osizeW_ + oh * osizeW_ + ow` (lines 285 and
360). With `channels_last_3d` `indices` those offsets land on unrelated
elements, so `atomicAdd`/`+=` scatters each `gradOutput` element to whatever
input position the misread index happens to name -- wrong gradients, no error.
The same applies to the destination: lines 550/574 take
`gradInput.mutable_data_ptr()` and lines 283/358 offset it as
`gradInput_ + d * isizeT_ * isizeH_ * isizeW_`, so a non-contiguous `gradInput`
is written through contiguous arithmetic too.

Both instances are silent and both are reachable from Python on the current
nightly (`2.15.0.dev20260920+xpu`): the `indices` one through
`torch.compile`, which saves `indices` in the input's memory format, and the
`gradInput` one through the `adaptive_max_pool3d_backward.grad_input` out=
overload -- pytorch's meta function requests contiguous strides with
`set_output_raw_strided(0, input.sizes(), {}, input.options())`
(`aten/src/ATen/native/AdaptiveMaxPooling3d.cpp:81`), and an empty `strides`
argument does not restride a caller-supplied `out`.

The XPU kernel is a direct port of the CUDA one, which carries the same
raw-`indices` pattern
(`aten/src/ATen/native/cuda/AdaptiveMaxPooling3d.cu:403,449,472`), so upstream
`pytorch/pytorch#197888` -- which scopes its fix to the CPU kernels because it
observes CUDA's compiled gradient as correct -- would not change the XPU path.
The sibling 2D kernel in this repository already carries the fix pattern:
`adaptive_max_pool2d_backward_kernel` normalizes `grad_output` and `indices` and
stages a contiguous `grad_input` with a copy-back
(`src/ATen/native/xpu/sycl/AdaptiveMaxPooling2dKernels.cpp:341-345,391-393`),
and `#3941` established that idiom for the non-contiguous-output case in the
same family. Fix strategy: apply that same normalization inside
`adaptive_max_pool3d_backward_kernel`.

- **Verdict:** IMPLEMENTING -- kernel-local fix, pattern already present in the sibling 2D kernel
- **Target repo:** torch-xpu-ops
- **Domain:** xpu-kernel

*Automated by fix-root-cause.*

</details>



<details>
<summary><b>Implementation</b></summary>

One file changed: `src/ATen/native/xpu/sycl/AdaptiveMaxPooling3dKernels.cpp`.
`adaptive_max_pool3d_backward_kernel` now normalizes `indices` the same way it
already normalized `gradOutput`, and stages a contiguous `gradInput` with a
copy-back when the caller's `gradInput` is not contiguous. Both dispatch lambdas
(atomic and non-atomic) take their pointers from the normalized tensors. This is
the idiom `adaptive_max_pool2d_backward_kernel` already uses in this repository.

```diff
 void adaptive_max_pool3d_backward_kernel(
     const Tensor& input,
     const Tensor& indices,
     const Tensor& gradInput) {
+  // Both kernel templates address gradOutput, indices and gradInput with
+  // contiguous index arithmetic, so all three have to be normalized here.
   const Tensor gradOutput_ = gradOutput.contiguous();
+  const Tensor indices_ = indices.contiguous();
+  const Tensor gradInput_ = gradInput.is_contiguous()
+      ? gradInput
+      : at::empty(gradInput.sizes(), gradInput.options());

-  gradInput.zero_();
+  gradInput_.zero_();
@@ (both AT_DISPATCH_FLOATING_TYPES_AND2 lambdas)
-          scalar_t* gradInput_data = gradInput.mutable_data_ptr<scalar_t>();
+          scalar_t* gradInput_data = gradInput_.mutable_data_ptr<scalar_t>();
           const scalar_t* gradOutput_data =
               gradOutput_.const_data_ptr<scalar_t>();
-          const int64_t* indices_data = indices.const_data_ptr<int64_t>();
+          const int64_t* indices_data = indices_.const_data_ptr<int64_t>();
@@ (end of function)
   }
+
+  if (!gradInput.is_contiguous()) {
+    gradInput.copy_(gradInput_);
+  }
 }
```

The alternative -- teaching the kernels to take strides, as the forward kernel
does for its input -- was rejected: the forward kernel encodes `argmax` as a
contiguous linear offset, so honoring arbitrary `gradInput` strides would mean
changing an index encoding that the CPU and CUDA kernels share.

Skip-guard review: APPROVE (no skip-shaped workaround; kernel-only change).

*Automated by fix-implement.*

</details>



<details>
<summary><b>Verification</b></summary>

Source build of pytorch `00ec4ca3a9` with this branch checked out at
`third_party/torch-xpu-ops` (`USE_XPU=1`, `TORCH_XPU_ARCH_LIST=bmg,pvc`), on an
Intel Arc Pro B60. `torch 2.15.0a0+git00ec4ca` imported from the source tree,
worktree clean.

| Case | Before (nightly `2.15.0.dev20260920+xpu`) | After |
|---|---|---|
| backward kernel, non-atomic template, `channels_last_3d` `indices` | wrong gradient | matches contiguous |
| backward kernel, atomic template, `channels_last_3d` `indices` | wrong gradient | matches contiguous |
| `torch.compile` (inductor) vs eager, `channels_last_3d` input | `max_diff=118` | `max_diff=0` |
| `torch.compile` (inductor) vs eager, contiguous input | `max_diff=0` | `max_diff=0` |
| `adaptive_max_pool3d_backward.grad_input` out= with `channels_last_3d` `gradInput` | wrong gradient | matches contiguous |

`aot_eager` matched eager both before and after, which is consistent with the
root cause being in the XPU kernel rather than in the decomposition.

```
pytest -q test/xpu/test_ops_xpu.py -k adaptive_max_pool3d       # 19 passed
pytest -q test/xpu/test_modules_xpu.py -k AdaptiveMaxPool3d     # 24 passed, 2 skipped
lintrunner -a -m HEAD~1                                        # no lint issues
```

- **Verdict:** PASSED

*Automated by fix-verify.*

</details>



## Summary

`adaptive_max_pool3d` backward on XPU indexed `indices` (and `gradInput`) with
contiguous arithmetic while only `gradOutput` was normalized, so a
`channels_last_3d` `indices` -- which is exactly what inductor saves for a
`channels_last_3d` input -- made the kernel scatter each gradient element to the
wrong input position. The fix normalizes `indices` and stages a contiguous
`gradInput` with a copy-back in
`src/ATen/native/xpu/sycl/AdaptiveMaxPooling3dKernels.cpp`, the same idiom the
sibling 2D backward kernel already uses. One file, +14/-4.

Verified on a source build on an Intel Arc Pro B60: the reporter's
eager-vs-inductor gap closes (`max_diff` 118 to 0), both backward kernel
templates now agree with the contiguous case, and the
`adaptive_max_pool3d_backward.grad_input` out= path is fixed as well.

**Follow-up (not fixed here):** the forward `adaptive_max_pool3d_kernel` has the
same defect for an `output`/`indices` pair supplied through the
`adaptive_max_pool3d.out` overload. It is a separate function, is not reachable
through `torch.compile`, and was left out to keep this change scoped to the
reported bug. Upstream CUDA carries the identical raw-`indices` pattern in both
directions, so it is worth reporting there too.

- **Verdict:** PASSED
- **Branch:** `agent/fix-issue-5475` (`ce3d7e3d`, base `6a2d048e`)
- Not pushed; the fix job exports the patch for human review.

_Generated by [fix job](https://github.com/intel/torch-xpu-ops/actions/runs/35684273723)._




<details>
<summary><b>Review</b></summary>

Copy into a new comment, keep ONE verdict, add a note. Two verdicts in
one comment are dropped as ambiguous.

```
**strong-accepted** <!-- review: strong-accepted -->
**weak-accepted** <!-- review: weak-accepted -->
**weak-rejected** <!-- review: weak-rejected -->
**strong-rejected** <!-- review: strong-rejected -->
**issue-drop** <!-- review: issue-drop -->

notes:
```

On a batch run add `item=<pytorch issue>` to say which sub-item the
verdict is about, **in the bold text and in the marker** -- the marker
is invisible, so the bold line is all a reader sees. One comment per
sub-item, since a bare verdict on a batch does not say which test it
means and is dropped. `run=<id>` only when judging a run other than the
newest, in both places too.

A CI DISABLED sub-item, `item=` from the fan-out table:

```
**strong-accepted item=197334** <!-- review: strong-accepted item=197334 -->
notes: same TensorAssign overload as upstream CUDA, fp8 dispatch list matches
```

Another sub-item of the same batch, judged on an earlier run:

```
**weak-rejected run=35171979678 item=197335** <!-- review: weak-rejected run=35171979678 item=197335 -->
notes: green only because the test is skipped, the kernel still lacks fp8
```

An `[xpu-alignment]` issue that should not have been filed -- one issue
per difference, so no `item=`:

```
**issue-drop** <!-- review: issue-drop -->
notes: accepted XPU/CUDA numerics difference, nothing to repair
```

</details>
